forum/communications/signals.py

```diff
@@ -0,0 +1,46 @@
+from django.db.models.signals import post_save
+from django.dispatch import receiver
+from channels.layers import get_channel_layer
+from asgiref.sync import async_to_sync
+from django.utils import timezone
+
+from .models import Message, Notification
+
+
+@receiver(post_save, sender=Message)
+def send_notification_via_channels(sender, instance, created, **kwargs):
+    """
+    Create a notification and broadcast it through Channels when a new message is sent.
+
+    Args:
+        sender (Model): The model class that sent the signal (Message).
+        instance (Message): The actual instance of the message that was created.
+        created (bool): A boolean indicating whether the instance was created.
+    """
+    if created:
+        room = instance.room
+        if room:
+            for participant in room.participants:
+                if participant.user_id != instance.sender.user_id:
+                    Notification.objects.create(
+                        user=participant,
+                        message=instance,
+                        is_read=False,
+                        created_at=timezone.now()
+                    )
+
+                    notification_data = {
+                        'user': participant.username,
+                        'message': instance.message,
+                        'is_read': False,
+                    }
+
+                    channel_layer = get_channel_layer()
+
+                    async_to_sync(channel_layer.group_send)(
+                        f'chat_{room.id}',
+                        {
+                            'type': 'send_notification',
+                            'notification': notification_data,
+                        }
+                    )
```

Comment:
Using `async_to_sync` inside the signal could block the main thread and cause performance bottlenecks.
You may use Django Channels' asynchronous features directly where possible or offload notifications to a Celery task.

Reply:
post method is coming from APIView and this is synchronous function.

If i try to await 

`await channel_layer.group_send( f"{str(receiver.user_id)}", {
                    "type": "notification.message",
                    "notification": NotificationSerializer(notification).data})` 

I get,

`django.core.exceptions.SynchronousOnlyOperation: You cannot call this from an async context - use a thread or sync_to_async.
`

Reply:
If `sync_to_async(channel_layer.group_send)(
                f"chat_{str(conversation_id)}", {"type": "chat.message", "message": MessageSerializer(message).data}
            )`

Error occurred: sync_to_async can only be applied to sync functions.

Reply:
About celery. Never used. I need more time to study what is Celery and how to use it.